filewatcher/v2/filewatcher.go

```diff
@@ -0,0 +1,567 @@
+//
+// Defaut to DefaultFSEventsDelay.
+func WithFSEventsDelay(delay time.Duration) Option {
+	return func(o *opts) {
+		o.fsEventsDelay = delay
+	}
+}
+
+// WithPollingInterval sets the interval to check file changes proactively.
+//
+// Default to DefaultPollingInterval.
+// To disable polling completely, set it to a negative value.
+//
+// Without polling filewatcher relies solely on fs events from the parent
+// directory. This works for most cases but will not work in the cases that
+// the parent directory will be remount upon change
+// (for example, k8s ConfigMap).
+func WithPollingInterval(interval time.Duration) Option {
+	return func(o *opts) {
+		o.pollingInterval = interval
+	}
+}
+
+// WithInitialReadInterval sets the interval to keep retrying to open the file
+// when creating a new FileWatcher, when the file was not initially available.
+//
+// Default to DefaultInitialReadInterval.
+func WithInitialReadInterval(interval time.Duration) Option {
+	return func(o *opts) {
+		o.initialReadInterval = interval
+	}
+}
+
+// WithFileSizeLimit sets the soft file size limit, with the hard limit being
+// 10x (see HardLimitMultiplier) of the set soft limit.
+//
+// This is completely ignored when DirParser is used.
+//
+// If the soft limit is violated,
+// the violation will be reported via slog at error level and prometheus
+// counter of limitopen_softlimit_violation_total,
+// but it does not stop the normal parsing process.
+//
+// If the hard limit is violated,
+// The loading of the file will fail immediately.
+//
+// Default to DefaultMaxFileSize.
+func WithFileSizeLimit(limit int64) Option {
+	return func(o *opts) {
+		o.fileSizeLimit = limit
+	}
+}
+
+// New creates a new FileWatcher.
+//
+// If the path is not available at the time of calling,
+// it blocks until the file becomes available, or context is cancelled,
+// whichever comes first.
+//
+// When logger is non-nil, it will be used to log errors,
+// either returned by parser or by the underlying file system watcher.
+// Please note that this does not include errors returned by the first parser
+// call, which will be returned directly.
+func New[T any](ctx context.Context, path string, parser Parser[T], options ...Option) (*Result[T], error) {
+	opt := opts{
+		fsEventsDelay:       DefaultFSEventsDelay,
+		pollingInterval:     DefaultPollingInterval,
+		initialReadInterval: DefaultInitialReadInterval,
+
+		fileSizeLimit: DefaultMaxFileSize,
+	}
+	for _, o := range options {
+		o(&opt)
+	}
+	hardLimit := opt.fileSizeLimit * HardLimitMultiplier
+
+	var data T
+	var mtime time.Time
+	var files []string
+
+	var lastErr error
+	for {
+		select {
+		default:
+		case <-ctx.Done():
+			return nil, fmt.Errorf(
+				"filewatcher: context canceled while waiting for file(s) under %q to load: %w, last err: %w",
+				path,
+				ctx.Err(),
+				lastErr,
+			)
+		}
+
+		var err error
+		data, mtime, files, err = openAndParse(path, parser, opt.fileSizeLimit, hardLimit)
```

Comment:
risk of [leaking undefined behavior](https://reddit.atlassian.net/wiki/spaces/BAS/pages/2355396743/Go+Style+Notes#Leaking-Undefined-Behavior)